### PR TITLE
Add load for argument matrix subscript lowering

### LINKED ISSUE #2958: Accessing matrix input data in hull shader's patch constant function results in a compiler error
### Title
Accessing matrix input data in hull shader's patch constant function results in a compiler error

### Functional impact
Requires shaders that compiled with FXC to be rewritten to workaround this compiler error

### Minimal repro steps
*What is the smallest, simplest set of steps to reproduce the issue. If needed, provide a project that demonstrates the issue.*  

1. Using this input file ([hull.txt](https://github.com/microsoft/DirectXShaderCompiler/files/4748128/hull.txt)), run the following commands:
`dxc.exe /Od /Zpr /nologo /Ths_6_0 /EHullFunc /Fo"hullout.tmp" "hull.txt"`
`dxc.exe /Od /Zpr /nologo /Ths_6_0 /EHullUnpackedFunc /Fo"hullunpackedout.tmp" "hull.txt"`
`dxc.exe /Od /Zpr /nologo /Ths_6_0 /EHullSimpleFunc /Fo"hullsimpleout.tmp" "hull.txt"`

### Expected result
The compile succeeds for all entry points in the file.

### Actual result
The compile using `HullFunc` as the entry point returns 0x80AA0018 (`DXC_E_GENERAL_INTERNAL_ERROR`). The other entry points, which either omit the matrix usage or access it as separate float4s, compile successfully.

### PR DESCRIPTION
**Pull request by pow2clk**
Matrix lowering was not handling subscript oeprations when the matrix
was a shader input. copying the matrix to a temporary explicitly or
implicitly as part of a copy in function call worked around the problem.
When the argument copy was eliminated, this problem was exposed.

By adding a load for the indicated matrix for a subscript of a shader
input matrix, the lowering can procede using the lowered matrix from the
load.

Additionally, hull shaders with their uninlineable functions were
failing to detect the constanthull function as a graphics function so
the subscript lowering was being treated as if it were in a library.
Even when it got to signature lowering, there was no support for
lowering matrix loads.

The test for shaderism now tests the function attributes of the module's
entry function. The signature lowering for matrix loads is moved into a
function that is called by input lowering for both the constant hull and
entry functions.

Added general tests for matrix subscripts from different memory types as
well as specific tests for the argument passing problem in pixel and
domain shaders.

Fixes #2958

### CODE REVIEW

File: lib/HLSL/HLSignatureLower.cpp [hunk 1]
```diff
@@ -654,6 +655,96 @@ Value *replaceLdWithLdInput(Function *loadInput, LoadInst *ldInst,
...
+
+  Value *newVec =
+    HLMatrixLower::BuildVector(matElts[0]->getType(), matElts, LocalBuilder);
+  newVec = MatTy.emitLoweredMemToReg(newVec, LocalBuilder);
+
+  CI->replaceAllUsesWith(newVec);
+  CI->eraseFromParent();
+}
+
+
```
Comment (pow2clk) [inline lib/HLSL/HLSignatureLower.cpp:747]: Just a note that these are just moved to subfunctions without alteration to the code.

File: lib/HLSL/HLSignatureLower.cpp [hunk 2]
```diff
@@ -964,84 +1055,11 @@ void GenerateInputOutputUserCall(InputOutputAccessInfo &info, Value *undefVertex
...
-          for (unsigned c = 0; c < MatTy.getNumColumns(); c++) {
-            unsigned matIdx = MatTy.getRowMajorIndex(r, c);
-            Value *Elt = LocalBuilder.CreateExtractElement(Val, matIdx);
-            LocalBuilder.CreateCall(ldStFunc,
-              { OpArg, ID, rowIdx, columnConsts[c], Elt });
-          }
-        }
-      }
-      CI->eraseFromParent();
+      replaceMatStWithStOutputs(CI, matOp, ldStFunc, OpArg, ID, columnConsts, vertexOrPrimID, idxVal);
```
Comment (pow2clk) [inline lib/HLSL/HLSignatureLower.cpp:1062]: Moving this code wasn't strictly necessary as it's still only called from here, but I moved it for consistency with the loading replacement code. I think it makes the control flow here easier to read too.

File: lib/HLSL/HLMatrixLowerPass.cpp [hunk 3]
```diff
@@ -406,7 +406,7 @@ Value *HLMatrixLowerPass::tryGetLoweredPtrOperand(Value *Ptr, IRBuilder<> &Build
     RootPtr = GEP->getPointerOperand();
 
   Argument *Arg = dyn_cast<Argument>(RootPtr);
-  bool IsNonShaderArg = Arg != nullptr && !m_pHLModule->IsGraphicsShader(Arg->getParent());
+  bool IsNonShaderArg = Arg != nullptr && !m_pHLModule->IsGraphicsShader(m_pHLModule->GetEntryFunction());
```
Comment (tex3d) [inline lib/HLSL/HLMatrixLowerPass.cpp:409]: Use `IsEntryThatUsesSignatures(Arg->getParent())` instead.

File: tools/clang/test/HLSLFileCheck/hlsl/types/matrix/matrix_subscript.hlsl [hunk 4]
```diff
@@ -0,0 +1,64 @@
+// RUN: %dxc -DMIDX=1 -DVIDX=2 -T ps_6_0 %s | FileCheck %s
+// RUN: %dxc -DMIDX=i -DVIDX=2 -T ps_6_0 %s | FileCheck %s
+// RUN: %dxc -DMIDX=1 -DVIDX=j -T ps_6_0 %s | FileCheck %s
+// RUN: %dxc -DMIDX=i -DVIDX=j -T ps_6_0 %s | FileCheck %s
```
Comment (tex3d) [inline tools/clang/test/HLSLFileCheck/hlsl/types/matrix/matrix_subscript.hlsl:4]: Let's add extra run line(s) to try lib target (like lib_6_3), and put [shader("pixel")] in front of entry point.  Same with other tests as it makes sense.

**Pull request merged**